pink/barriers/barrier.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 Stéphane Caron
+
+"""All barriers derive from the :class:`Barrier` base class.
+
+The formalism used in this implementation is written down in
+examples/barriers/NOTES.md
+"""
+
+import abc
+from typing import Callable, Optional, Tuple, Union
+
+import numpy as np
+
+from ..configuration import Configuration
+
+
+class Barrier(abc.ABC):
+    r"""Abstract base class for barrier.
+
+    A barrier is a function :math:`\boldsymbol{h}(\boldsymbol{q})` that
+    satisfies the following condition:
+
+    .. math::
+
+        \frac{\partial \boldsymbol{h}_j}{\partial \boldsymbol{q}}
+        \dot{\boldsymbol{q}} +\alpha_j(\boldsymbol{h}_j(\boldsymbol{q}))
+        \geq 0, \quad \forall j
+
+    where :math:`\frac{\partial \boldsymbol{h}_j}{\partial \boldsymbol{q}}`
+    are the Jacobians of the constraint functions, :math:`\dot{\boldsymbol{q}}`
+    is the joint velocity vector, and :math:`\alpha_j` are extended
+    class K functions.
+
+    Attributes:
+        dim: Dimension of the barrier.
+        gain: barrier gain.
+        class_k_fn: Extended class K function.
```

Comment:
Can we find a name for this function that refers to what it does, rather than its properties?

It could help newcomers build their intuition. As a newcomer to CBFs too, looking at the formula what I gather is that $\alpha$ maps $h(q)$ (that lives in "barrier configuration space") to "barrier velocity space".

Reply:
Actually, the meaning of class kappa function is more simple. It is a generalisation of gain of $h(q)$ function, it allows to define influence of stabilisation term more flexibly. As an example, consider case of joint barrier, which I've defined in one of the further discussions.

As for naming, we suggest to rename it as `gain_function` to make its meaning more transparent.